### PR TITLE
llvm/16: Fix compiler-rt missing sanitizers when using useLLVM

### PR DESCRIPTION
**Pull request by s1341**
Attempts to address https://github.com/NixOS/nixpkgs/issues/97688 by refactoring the llvm/16 toolchain build to make more sense.

The build order is now:
1. {compiler-rt-no-libc}
2. {compiler-rt-no-libc, libc}
3. {compiler-rt-no-libc, libc, libc++}
4. {compiler-rt-no-libc, libc, libc++, compiler-rt-libc}

The names of the stages have been updated to reflect this.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by s1341** [comment]
Hrm. I don't know why it is infinite-recursioning. I don't have that issue locally. How can I reproduce what ofborg ran?

File: pkgs/development/compilers/llvm/16/default.nix
Comment (Artturin) [inline pkgs/development/compilers/llvm/16/default.nix:1]: Don't reformat the file, it needs to be similar to the other llvm versions.

It's impossible to tell what changes you've made in it.

**Review by SuperSandro2000 — COMMENTED** [review]
This PR rebuilds a lot of packages which means we must target staging. Please follow the [contributing guide](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#rebasing-between-branches-ie-from-master-to-staging) to not potentially ping a lot of people.

**Comment by SuperSandro2000** [comment]
Locking this to avoid more pings because of the bad rebase. Please open a new PR.

**Comment by Ericson2314** [comment]
https://github.com/NixOS/nixpkgs/pull/297144 is the successor PR.

**Pull request closed without merging**